Adding experiment variant entity

martin-gaievski opened the pull request:
### Description
Adding sub experiment entity to address hybrid search experiments. Items covered in this PR:
- SubExperiment DAO
- index mapping for sub-experiment
- SubExperiment model class
- functionality: created new entry in sub-experiment index for each option of hybrid search experiment. each item has valid set of options - normalization/combination/weights. It has experiment reference and status "PROCESSING"
- update sub-experiment entry as per actual status - COMPLETED/ERROR
- figure out format for storing results and link result to sub-experiment and all sub-experiments to experiment

Two indexes are affected by this change:

experiments (existing):
```
{
   id,
  ...
  result: {
       query_text: {
            search_configuration : {
                  sub_experiment_id1: evaludation_id1,   <-- this is added
                  sub_experiment_id2: evaludation_id2
            }
       }
  }
}
```
sub_experiment (new):
```
{
   id,
   timestamp,
   status,
   type,
   experiment_id,
   parameters,                                           <-- this is generic, we store options like normalization/combination techniques etc.
   results: {evaluation: evaludation_id }  <-- ref to evaluation record with exact metrics for this query and set of parameters
}
```


following is example of sub-experiment documents from new index:

```
GET /.plugins-search-relevance-sub-experiment/_search
{
    "hits": {
        "total": {
            "value": 120,
            "relation": "eq"
        },
        "max_score": 1.0,
        "hits": [
            {
                "_index": ".plugins-search-relevance-sub-experiment",
                "_id": "efaf1bdc-3db7-4b41-8272-aad3661c40c9",
                "_score": 1.0,
                "_source": {
                    "id": "efaf1bdc-3db7-4b41-8272-aad3661c40c9",
                    "timestamp": "2025-05-29T01:54:13.468Z",
                    "type": "HYBRID_SEARCH",
                    "status": "PROCESSING",
                    "experimentId": "71eef27f-3cba-4eac-aaad-a85dfa72e37c",
                    "parameters": {
                        "weights": "[0.9000001, 0.099999905]",
                        "normalization": "min_max",
                        "combination": "arithmetic_mean"
                    },
                    "results": {}
                }
            },
            {
                "_index": ".plugins-search-relevance-sub-experiment",
                "_id": "6ef37935-dc3e-496a-9514-73f931bdc24a",
                "_score": 1.0,
                "_source": {
                    "id": "6ef37935-dc3e-496a-9514-73f931bdc24a",
                    "timestamp": "2025-05-29T01:54:13.466Z",
                    "type": "HYBRID_SEARCH",
                    "status": "PROCESSING",
                    "experimentId": "71eef27f-3cba-4eac-aaad-a85dfa72e37c",
                    "parameters": {
                        "weights": "[0.8000001, 0.19999993]",
                        "normalization": "min_max",
                        "combination": "geometric_mean"
                    },
                    "results": {}
                }
            }, ...
            }
        ]
    }
}
```

and following is example of relevance experiment table with sub-experiment records
```
{
    "hits": {
        "total": {
            "value": 1,
            "relation": "eq"
        },
        "max_score": 1.0,
        "hits": [
            {
                "_index": ".plugins-search-relevance-experiment",
                "_id": "9c2af9ee-6b92-4ca5-9b55-64d5f07110d4",
                "_score": 1.0,
                "_source": {
                    "id": "9c2af9ee-6b92-4ca5-9b55-64d5f07110d4",
                    "timestamp": "2025-05-30T00:46:52.147Z",
                    "type": "HYBRID_SEARCH",
                    "status": "COMPLETED",
                    "querySetId": "721605b3-aa09-4a94-b0cf-b73887217d87",
                    "searchConfigurationList": [
                        "ef14a1e7-573b-4892-bba2-d10485709060",
                        "97f3d6e4-4542-4194-a670-c18f4a30d22d"
                    ],
                    "judgmentList": [
                        "8377132a-0422-4f12-9259-ae73ab05871f"
                    ],
                    "size": 2,
                    "results": {
                        "phone": {
                            "97f3d6e4-4542-4194-a670-c18f4a30d22d": {
                                "6ae66963-1dae-4083-b7b6-23bc7abcc687": "7569a44e-5b4f-4431-862c-ff693d799355",
                                "d7039f50-6977-4b67-a2db-6ebeeef940fa": "a8293c6a-88dc-452d-85f6-de5e9b1328e2",
                                "bf50b7ed-3894-4228-9c16-fcd0b6d711ad": "cb4d48df-f0f4-4dc8-8537-92104877b7b4",
                                "199f3081-9f44-4b97-8bb9-c222dcf39989": "089cce8b-ed1c-4668-8e61-8580593d9891",
                                "742b6323-8adb-4470-8eda-5b2e647e9b56": "5bc3edf3-cced-4114-bc92-46a783efcfd2",
                                "a774c97b-b86d-4f27-bf98-4b6ab80d1485": "703eaabf-1ab2-4d6c-b899-c5ddfd48073c",
                                "375bb54c-fa4e-4fe8-b0c2-773fb7be318e": "39fb4a5f-9faf-4d52-86d2-f9bcf5861039",
                                "e2a60549-ade8-495a-a3f3-83ca796dafde": "8e65e167-e15a-4abf-985c-028b1b871dbe",
                                "0e389fb0-671a-4c80-939e-286640ce3f9e": "0279681f-0820-4e86-a2cb-a19facb084a4",
                                "154d5c05-9920-4a3a-bf66-fe19a7f836ee": "fa130a60-80d0-4548-a544-a3eb92f670a0",
                                "d03a37ea-56db-4c0f-bac1-22e1d4079802": "18a3c99e-c84e-4506-98ba-bfd1b3f18110",
                                "17da7fda-2dac-40ec-994b-073865d2d810": "0dc8ce53-5455-447d-b4fb-2fbff3020927",
                                "23ff2ef9-2d25-4065-91db-9ade0519361e": "60308c15-68e1-47b9-ab22-2ae2e0dbacba",
                                "971385b3-0bd8-445f-95c1-7303eec4e633": "1a3f467a-9938-44c9-a6b3-560343730c90",
                                "316dbdca-f351-4074-9df4-078f5cb04ad0": "a1563daf-31a0-432d-9745-6ae1a533b315",
                                "2ae6ff31-a680-4c1a-9b04-f97aead9fba6": "77e10587-b9cc-4ec3-8bf2-59a9ff1a73a1",
                                "9a6e8d6e-4743-4c5c-90f2-19f5474465b6": "90a0ed8b-85e3-4521-b6e1-5f93782982fd",
                                "65a847d4-7628-4ad0-9781-5459069ce2e3": "bcb63ec3-b816-4c0d-88a4-5413185f4e68",
                                "8cadc5f4-fa44-468d-bce0-0cb3b1c789cf": "2e46eeee-185a-4054-a977-91939c666971",
                                "141e4301-d1d1-4873-b747-bd4cf1a5aee6": "e09a695a-c211-408f-bd7e-c8d007b6e0e7",
                                "cafa6497-82b7-40c7-b630-777fb87e50dc": "d09a7f91-8a34-41b0-8ff4-129d9232d69c",
                                "a13b8af6-b71b-49b9-a1e1-58f992c85aae": "d09a5f8b-4cee-4c43-b23d-1dfda62f03ff",
                                "8c398f28-a816-420a-9d4b-98423bce5afc": "047efba9-f125-4231-9442-85275025f8de",
                                "2ec34922-24e7-43c9-8c1d-bf7bb05e3cd2": "347ea896-8d1f-4373-b271-26e3c673608c",
                                "01739096-2504-4245-b6c7-6405ce14f719": "a5c28dd8-ab98-4f3e-b529-07cf44bd62b4",
                                "3e6c30c9-affe-48fb-aa2b-25205c3cdc4a": "a93f1603-a074-4c27-ab03-207a4c04c373",
                                "8679d187-f9d6-4a32-baa6-ae8177aaa641": "4c1ccee6-4200-47d2-b05c-cd9a600aa57d",
                                "55c197aa-5d4a-4122-a4e6-2d484f39c682": "f285b66f-50df-43ae-acf3-54965a4564ca",
                                "6bf36e7c-d8e4-4bae-8d3f-93c77a2fe5ad": "ff33a882-19ac-4753-be41-e4f7aded0623",
                                "2b7d9e7e-07c4-44f2-b460-a6c8565f05f9": "6ed6603b-68c5-405b-8c24-2fae44f9de96",
                                "bb934228-b55d-4b86-8f96-1142504a89ae": "272a9fec-29d9-490f-af6d-b08ff32d7dcc",
                                "0c47ffbb-fed3-4955-8f4d-a0cb40715a7e": "75a70bad-ba5b-4bd4-9d55-72c0db4402e9",
                                "d4cf3c45-1ee0-4bd8-a8fa-211b73e39bf0": "feaf9284-8ab1-498a-b98a-413b9fa7bea0",
                                "404522f3-011b-438a-9c2a-d0bf02ee5ebb": "e4b5e60b-5d45-497e-8e01-a32c48c19b8a",
                                "2a59bf5a-a0d0-4d82-bbb9-fc954e2e1ba7": "ca0e7b73-5970-41c8-a484-d0cfabf4eedf",
                                "c6decd3d-4267-4a70-abe2-292fb5d1b924": "c0118a7f-03f5-44ff-bcc2-930c47ef4e81",
                                "7a648bc9-f12c-422e-b181-159725825151": "4f410dd4-939b-4b29-85d3-05e89a87f89a",
                                "a89e6270-a3f3-4d73-8722-6ec88291f761": "9bcee00a-83a0-43da-9fc2-25802143d591",
                                "68a027f9-4e26-4e06-8539-6f0df1ad1f43": "4a6b8db0-bad8-4e5d-845a-1fbe7c37e733",
                                "63a30c9f-48c5-4bd0-a373-5888e2319ed7": "da68ea3b-799f-4cf2-abdf-3471b7fe27ab",
                                "529a2601-a429-4335-bce0-01a076c75211": "bfdf98b4-14bf-4198-8a34-c1bc794bfccf",
                                "9dfb1859-f6c5-4387-b3c7-fd23931a8f7f": "527d7d04-d6b7-4532-994a-f5a473c1b235",
                                "ec1441bf-5d3e-4219-bc72-149612fcc86f": "7797a046-979e-48ee-8f59-4ddd3329a2f6",
                                "b0b3e41d-5f0a-4d1e-bff4-bcb98ded5d2b": "21781949-d7a0-4a1b-95c2-aa49b2326477",
                                "936546b7-8361-48e7-8d5e-37d23de5d604": "46bc7410-35db-44eb-95e7-5ffe0767ac65",
                                "b29af3c5-ed98-432e-b925-315d3e71c502": "dd8c7320-e53f-473a-8122-2dce2ad4b2d4",
                                "5ebee4b3-80be-4bd4-b682-4eaab3414fcf": "452c13a1-357c-46ff-867a-0341d686831f",
                                "4c9d9e4e-56c7-483e-a5e8-1ed9828db596": "33a55d96-84eb-4196-b90f-001a7e0df019",
                                "954eb379-2021-41f0-8eed-743b8ed8db49": "471a174a-a4ea-4ca7-8251-80f6e3719d9a",
                                "4e6ccdf4-9aad-416a-b7e1-bcb70926ab98": "355dae9c-08d8-4eda-9aa2-9ab332af02f5",
                                "a88940d6-0460-4aae-9440-6fd535dfc656": "cb437e72-7e7d-467f-ac1d-99603571155a",
                                "b2f672f4-4143-4143-b742-000f506bcf3d": "67eff083-3ecc-4b5e-862e-e04225032a3a",
                                "ee9724f5-f4a5-4fa0-aab2-e72280f93e39": "49ae2d2d-00e0-4beb-8843-69b3004abebf",
                                "eb6a495a-12a6-4190-a974-255968b474eb": "94186c76-7772-49ba-aebc-b380f05d4528",
                                "aa232134-4c4e-47a7-9dac-2584fb4a2e77": "6706b476-496f-40ed-8005-6b5a525815b3",
                                "515696e0-50bd-4340-94a5-76b17d5921d4": "531e55ce-d3d0-4e49-bd2b-52038e6a3802",
                                "3d2d4662-645d-4d82-8b22-71c1d651877a": "4c775cbc-6c4d-42f2-8678-da27136da747",
                                "aaf46576-cab7-4d01-a2dc-0e9e01056e62": "94b544be-7835-4646-91ca-27cd85bd5c76",
                                "10ffb7d3-c550-4916-a855-0b4444ea7f59": "306baecf-8bdf-4bc1-a1a9-becdeb67d9bd",
                                "71cbdfe6-bdea-482c-a5bf-a9d923e1e0d3": "c612904b-7581-4e1a-997a-c59d56686925"
                            },
                            "ef14a1e7-573b-4892-bba2-d10485709060": {
                                "6ae66963-1dae-4083-b7b6-23bc7abcc687": "76693c30-36fb-463c-8259-319f1aaacbf4",
                                "d7039f50-6977-4b67-a2db-6ebeeef940fa": "942d26a9-e00c-4413-b19f-7203ecde3bb6",
                                "bf50b7ed-3894-4228-9c16-fcd0b6d711ad": "817862f9-be6b-49c8-ae56-84d8410f7bda",
                                "199f3081-9f44-4b97-8bb9-c222dcf39989": "50866924-36b5-4da9-9eca-98b87f955e03",
                                "742b6323-8adb-4470-8eda-5b2e647e9b56": "64b64d5a-5281-4680-a172-c47506a8fe3a",
                                "a774c97b-b86d-4f27-bf98-4b6ab80d1485": "7476c68e-420b-4668-9e54-0516704fe164",
                                "375bb54c-fa4e-4fe8-b0c2-773fb7be318e": "2bc2d99e-dbf4-428a-a347-bafc7eccbdd4",
                                "e2a60549-ade8-495a-a3f3-83ca796dafde": "a5211495-21c2-431a-98e6-190070daf505",
                                "0e389fb0-671a-4c80-939e-286640ce3f9e": "c6f7fd7a-ad0f-4648-8d3c-158a72a8d2c7",
                                "154d5c05-9920-4a3a-bf66-fe19a7f836ee": "4d40da7f-cfdf-481e-b2bc-1b6185c34b78",
                                "d03a37ea-56db-4c0f-bac1-22e1d4079802": "efa62cd2-4c56-48cf-ba73-2ced53790897",
                                "17da7fda-2dac-40ec-994b-073865d2d810": "9cc8e9ca-cdbd-4256-813c-0f37976312d4",
                                "23ff2ef9-2d25-4065-91db-9ade0519361e": "22ef6d4b-b892-49ac-ae3d-17ef1a8647ff",
                                "316dbdca-f351-4074-9df4-078f5cb04ad0": "dc902e4a-a5f3-40c9-92a0-689217e8d097",
                                "971385b3-0bd8-445f-95c1-7303eec4e633": "354c429c-ed07-4a93-8301-df15d822f0a5",
                                "2ae6ff31-a680-4c1a-9b04-f97aead9fba6": "3fa6da47-c552-41ef-bd12-3b3fe3d2ae99",
                                "9a6e8d6e-4743-4c5c-90f2-19f5474465b6": "77bbe8f8-fb93-4e22-8360-589f15497753",
                                "8cadc5f4-fa44-468d-bce0-0cb3b1c789cf": "4f120130-20ab-450c-bdbe-386c4d213315",
                                "141e4301-d1d1-4873-b747-bd4cf1a5aee6": "08af4f6d-331f-4098-99a6-2c3ad34b9138",
                                "65a847d4-7628-4ad0-9781-5459069ce2e3": "7e75c505-0394-483d-a47a-bc85824cd1f9",
                                "a13b8af6-b71b-49b9-a1e1-58f992c85aae": "652a2ea7-7fca-4c2f-b72b-4f559ca1eb84",
                                "cafa6497-82b7-40c7-b630-777fb87e50dc": "3cefaf18-815f-4b73-a681-974acc6cb0dd",
                                "8c398f28-a816-420a-9d4b-98423bce5afc": "0c9b24d8-66c7-4c02-9d1b-c0ad0ce405c2",
                                "2ec34922-24e7-43c9-8c1d-bf7bb05e3cd2": "d4c6d0ce-4110-4343-9d34-05bda7830792",
                                "01739096-2504-4245-b6c7-6405ce14f719": "255c901e-e2a5-462f-adba-9418741adf35",
                                "3e6c30c9-affe-48fb-aa2b-25205c3cdc4a": "a02630c3-ea88-4e81-8add-296279afce7b",
                                "8679d187-f9d6-4a32-baa6-ae8177aaa641": "2adf9b39-0719-476e-96c0-452dd256b4de",
                                "55c197aa-5d4a-4122-a4e6-2d484f39c682": "12019bf3-42b1-48ab-bff7-fad1611a27b5",
                                "6bf36e7c-d8e4-4bae-8d3f-93c77a2fe5ad": "54c14f36-0be6-49ab-ad8e-b10ffbd68182",
                                "2b7d9e7e-07c4-44f2-b460-a6c8565f05f9": "cc48fede-48e0-4da6-adbc-29a9021c97f3",
                                "bb934228-b55d-4b86-8f96-1142504a89ae": "17e79f4a-d051-4f6d-b9f1-e2a67885baba",
                                "0c47ffbb-fed3-4955-8f4d-a0cb40715a7e": "61688b02-63be-4672-b806-b6f356831c05",
                                "d4cf3c45-1ee0-4bd8-a8fa-211b73e39bf0": "03ca69ff-d000-4059-bd85-e4690be925f8",
                                "404522f3-011b-438a-9c2a-d0bf02ee5ebb": "354a0d7a-ee96-4d3b-9fcb-16136af404d5",
                                "2a59bf5a-a0d0-4d82-bbb9-fc954e2e1ba7": "d997e911-fc26-48d1-bfee-805b21672927",
                                "c6decd3d-4267-4a70-abe2-292fb5d1b924": "c3e2a39b-d608-433c-b6e3-910ac93e204b",
                                "7a648bc9-f12c-422e-b181-159725825151": "a6ff2abb-4b38-4548-927e-ff98f37988db",
                                "a89e6270-a3f3-4d73-8722-6ec88291f761": "cec3a913-4f58-4d5c-a40b-418bca088ce5",
                                "63a30c9f-48c5-4bd0-a373-5888e2319ed7": "f6e10a9c-5e53-4ecf-9674-d133a5776602",
                                "68a027f9-4e26-4e06-8539-6f0df1ad1f43": "2a4e0beb-35ed-4f80-b6ba-68cd60d5b23c",
                                "529a2601-a429-4335-bce0-01a076c75211": "1f08fcc0-51f3-4b7d-9236-cbca4ebb5a69",
                                "9dfb1859-f6c5-4387-b3c7-fd23931a8f7f": "4744aecb-bf86-4780-ae30-81dad9a85223",
                                "ec1441bf-5d3e-4219-bc72-149612fcc86f": "b5507b96-b985-4072-a593-f01d45ad249f",
                                "b0b3e41d-5f0a-4d1e-bff4-bcb98ded5d2b": "80b1f2a7-f39e-4216-af0a-d9023811ed42",
                                "936546b7-8361-48e7-8d5e-37d23de5d604": "1e98724b-3ac2-425d-b378-ea0f03294136",
                                "b29af3c5-ed98-432e-b925-315d3e71c502": "12b403ea-3581-41e4-8713-952160db297f",
                                "5ebee4b3-80be-4bd4-b682-4eaab3414fcf": "322fdf2b-010a-42bd-ad3a-f44b081b469f",
                                "4c9d9e4e-56c7-483e-a5e8-1ed9828db596": "a2f36f71-6523-4fce-80e0-c25bf88e8fc8",
                                "954eb379-2021-41f0-8eed-743b8ed8db49": "27a99128-f4ca-44e4-acfd-75bd8cfb0bd5",
                                "4e6ccdf4-9aad-416a-b7e1-bcb70926ab98": "b211b1bc-a9f8-4ed0-a1e5-62eae2bd4a6b",
                                "a88940d6-0460-4aae-9440-6fd535dfc656": "ef292f15-b21b-4b15-94f1-18c731e1694b",
                                "b2f672f4-4143-4143-b742-000f506bcf3d": "aa164fd3-1b0d-42a0-83a5-15a9b5ad0c17",
                                "ee9724f5-f4a5-4fa0-aab2-e72280f93e39": "5c8ae547-ab16-428d-9070-860c328e46ac",
                                "eb6a495a-12a6-4190-a974-255968b474eb": "7d1aa37b-3283-443e-9dcf-023775a2aa2e",
                                "aa232134-4c4e-47a7-9dac-2584fb4a2e77": "349911fc-7c49-428a-a3fa-994dbe5ed4aa",
                                "515696e0-50bd-4340-94a5-76b17d5921d4": "fe6d5f68-0921-41cc-81bd-7b1358cb9323",
                                "3d2d4662-645d-4d82-8b22-71c1d651877a": "ce895509-d277-44b4-97fb-0444a1918b2d",
                                "aaf46576-cab7-4d01-a2dc-0e9e01056e62": "0dfa0b9b-69f7-4394-bc0c-15bb8abb68cb",
                                "10ffb7d3-c550-4916-a855-0b4444ea7f59": "a94757d4-1708-4da1-9451-98e948354759",
                                "71cbdfe6-bdea-482c-a5bf-a9d923e1e0d3": "408d58d4-aa64-4cae-94b2-ccf3788ffe7c"
                            }
                        },
```



future work:
- add API to retrieve evaluations to later show them in UI 

### Issues Resolved
https://github.com/opensearch-project/search-relevance/issues/29

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

fen-qin reviewed (commented):
Overall looks good to me. One concern is the compatibility with frontend experiment view.
- Currently experiment view will read the experiment by experiment_id, then read the evaluation results. the access to experiment if by a GET api, and the access to evaluation results is by a direct SEARCH against evaluation results, which is not a system index
- The implementation put sub experiment as one of the system indices. it's hard to read by SEARCH if we protect our system indices with security plugin. I would suggest either move it out of system indices or implement a GET API to access the data.

martin-gaievski commented:
> > Overall looks good to me. One concern is the compatibility with frontend experiment view.
> > 
> > * Currently experiment view will read the experiment by experiment_id, then read the evaluation results. the access to experiment if by a GET api, and the access to evaluation results is by a direct SEARCH against evaluation results, which is not a system index
> > * The implementation put sub experiment as one of the system indices. it's hard to read by SEARCH if we protect our system indices with security plugin. I would suggest either move it out of system indices or implement a GET API to access the data.
> 
> I think with sub-experiments we should follow same pattern as in existing Experiment entity, this means use the protected index. Eventually the sub-experiment is an extension of experiment with more parameters, it can even be part of the experiment itself. In this case I assume I need rest action class and route path for GET in sub_experiment, similar to experiment GET:
> 
> * [action](https://github.com/opensearch-project/search-relevance/blob/main/src/main/java/org/opensearch/searchrelevance/rest/RestGetExperimentAction.java)
> * [route](https://github.com/opensearch-project/search-relevance/blob/main/src/main/java/org/opensearch/searchrelevance/rest/RestGetExperimentAction.java#L47-L52) and [path](https://github.com/opensearch-project/search-relevance/blob/main/src/main/java/org/opensearch/searchrelevance/common/PluginConstants.java#L24)
> 
> is this something you had in mind @fen-qin ?

based on the offline discussion, I changed sub-experiment items from protected index to open/user visible index

martin-gaievski commented:
We reviewed this PR offline with @epugh, these are AI for me:
- change type name for experiment from HYBRID_SEARCH to HYBRID_OPTIMIZER
- came up with better name instead of Sub Experiment

martin-gaievski commented:
With recent comment https://github.com/opensearch-project/search-relevance/pull/31/commits/b2543711d5e4b3a4e0cd94ce627c01f2dad8d3fd I pushed following changes:
- renamed Sub Experiment to Experiment Variant
- renamed experiment type to HYBRID_OPTIMIZER
